packages/kbn-securitysolution-utils/src/client_concurrency/index.ts

```diff
@@ -0,0 +1,57 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+import limit from 'p-limit';
+
+/**
+ * Helper function to call an API multiple times concurrently, basically like a client-side bulk API.
+ * 
+ * const responses = await clientConcurrency(
+        detectionsClient.createRule.bind(detectionsClient),
+        ruleCopies.map((rule) => ({
+          body: rule,
+        }))
+      );
+ * Note the `.bind` is crucial in `detectionsClient.createRule.bind(detectionsClient)` - without binding, `createRule` is called without
+      access to the detectionsClient object and it will fail at runtime - even though the code typechecks fine.
+ * @param clientFunction The API function to call multiple times
+ * @param inputs An array of input values. Each value in the array is passed to a separate invocation of the clientFunction.
+ * @param concurrency Number of simultaneous in-progress API calls that are allowed.
+ * @returns An array of the responses from the API calls.
+ */
+export const clientConcurrency = async <Input, Output>(
+  clientFunction: (input: Input) => Promise<Output>,
+  inputs: Input[],
+  concurrency: number = 10
+) => {
+  const limiter = limit(concurrency);
+  const promises = inputs.map((input) => limiter(() => clientFunction(input)));
+  return Promise.all(promises);
+};
```

Comment:
This function is an alternative way of implementing concurrency compared to the one below (`concurrentlyExec`). IMO `concurrentlyExec` is a bit better because it's easy to forget the `.bind` in this implementation, but `concurrentlyExec` does require creating a wrapper function for each request which makes the code look a bit messier (but the compiler will help you out, unlike with `.bind` where the compiler does not help you).

Reply:
I'll remove one of the concurrency implementations depending on which one reviewers think is better DX-wise.

Reply:
IMHO `concurrentlyExec` has much more straightforward API. It's quite easy to forget adding `.bind()`. Though I think params should be renamed for better clarity. In fact `concurrentlyExec` accepts factories (factory is a well known and convenient pattern) so it should be reflected clearly like in the example below

```ts
type ActionFactory = () => Promise<Output>;

export const concurrentlyExec = async <Output>(
  actionFactories: ActionFactor[],
  concurrency: number = 10
) => {
  // ...
}
```